### PR TITLE
Removed Container, Linker and other classes from theano package

### PR DESCRIPTION
**Pull request by ionuttamas**
Fixes comments made on #233.

### CODE REVIEW

File: theano/__init__.py [hunk 1]
```diff
@@ -114,7 +114,6 @@ def disable_log_handler(logger=theano_logger, handler=logging_default_handler):
     utils,
 )
 from theano.gradient import Lop, Rop, grad, subgraph_grad
-from theano.link.basic import Container, Linker, LocalLinker, PerformLinker
 from theano.link.c.basic import CLinker, DualLinker, OpWiseCLinker
```
Comment (twiecki) [inline theano/__init__.py:117]:
```suggestion
```

**Comment by ionuttamas** [comment]
Failing unit tests, fail both on master as well as on this branch:

image

**Comment by michaelosthege** [comment]
Hi @ionuttamas ,
Try to rebase your branch on latest master. Our tests should be working fine - however I typically run them with `pytest tests/ -v -s` directly from the command line.

When it comes to kicking things from the main `__init__` we should take out even more: All of the linkers, but also things like `_as_array`, `object` and `object2`, `SymbolicInput/SymbolicOutput`...

**Comment by ionuttamas** [comment]
> Hi @ionuttamas ,
> Try to rebase your branch on latest master. Our tests should be working fine - however I typically run them with `pytest tests/ -v -s` directly from the command line.
> 
> When it comes to kicking things from the main `__init__` we should take out even more: All of the linkers, but also things like `_as_array`, `object` and `object2`, `SymbolicInput/SymbolicOutput`...

The failing tests (I think) are a problem on my machine, as running the latest master still makes them fail (and the same 7 tests always).
I am using Windows Subsystem for Linux that may cause some incompatibilities (I get: "ImportError: /home/iotama/.theano/compiledir_Linux-4.19-microsoft-standard-x86_64-with-glibc2.29-x86_64-3.8.5-64/tmp2avx4e28/ma62c58fa0e1eb04ad8e5318ad85066e7fb0233a39b4c6dd1c4d0270ae55fcede.so: undefined symbol: dgemm_

../theano/link/c/cmodule.py:297: ImportError").

I have also removed the above mentioned dependencies (_asarray, object2, as_op SymbolicInput/SymbolicOutput). 
I do get these steps failures due to code styles (and I'm not very sure what I'm missing wrong here ...):

- Tests / Check code style (pull_request)
- Tests / All tests (pull_request)

**Comment by michaelosthege** [comment]
> > Hi @ionuttamas ,
> > Try to rebase your branch on latest master. Our tests should be working fine - however I typically run them with `pytest tests/ -v -s` directly from the command line.
> > When it comes to kicking things from the main `__init__` we should take out even more: All of the linkers, but also things like `_as_array`, `object` and `object2`, `SymbolicInput/SymbolicOutput`...
> 
> @michaelosthege
> The failing tests are (I think) a problem on my machine, as running the latest master still makes them fail (and the same 7 tests always fail).
> I am using Windows Subsystem for Linux that may cause some incompatibilities (I get: "ImportError: /home/iotama/.theano/compiledir_Linux-4.19-microsoft-standard-x86_64-with-glibc2.29-x86_64-3.8.5-64/tmp2avx4e28/ma62c58fa0e1eb04ad8e5318ad85066e7fb0233a39b4c6dd1c4d0270ae55fcede.so: undefined symbol: dgemm_
> ../theano/link/c/cmodule.py:297: ImportError").
> 
> I have also removed the above mentioned dependencies (_asarray, object2, as_op SymbolicInput/SymbolicOutput).
> 
> I do get these steps failures **due to code styles** (and I'm not very sure what I'm missing here ...):
> 
> * Tests / Check code style (pull_request)
> * Tests / All tests (pull_request)

For the code style stuff you can set up pre-commit as described in https://github.com/pymc-devs/pymc3/wiki/PyMC3-Python-Code-Style
If you have the repo on another drive than C:\ you might run into a problem. In this case just create a `precommit.bat` file with the following content and run it manually before committing:

```
black theano/ tests/
isort theano/ tests/
flake8 theano/ tests/
```

I don't run it with the WSL, but I can imagine that it might give some problems. But everything also works fine on Windows also without WSL.
You can peek some conda install commands from https://gist.github.com/michaelosthege/b95d57ae8817eb1cfd19aa30beac2da6

**Comment by ionuttamas** [comment]
@michaelosthege some progress seem to be made, but getting stuck at figuring out: 
 - codecov/patch: Failing after 1s — 90.90% of diff hit (target 100.00%)   

Is this due to low unit tests code coverage?

**Comment by michaelosthege** [comment]
> @michaelosthege some progress seem to be made, but getting stuck at figuring out:
> 
> * codecov/patch: Failing after 1s — 90.90% of diff hit (target 100.00%)
> 
> Is this due to low unit tests code coverage?

Yes, it's due to low general coverage.
But your PR is more of a refactor and does not add new code, so this is fine.
I think after another rebase on latest master, and ideally also combining the two commits into one, we're ready to merge.

**Review by twiecki — DISMISSED** [review]

**Comment by michaelosthege** [comment]
Looks like I don't have the permission to merge unless the coverage check is also green.

**Comment by twiecki** [comment]
Thanks @ionuttamas!

**Pull request merged**